src/mos.c

```diff
@@ -906,48 +1142,349 @@ UINT24	mos_CD(char *path) {
+			pattern = mos_strdup(inputPath + 1);
+			if (!pattern) goto cleanup;
+			usePattern = TRUE;		
+		}
+    } else {
+        char *lastSeparator = strrchr(inputPath, '/');
+        if (lastSeparator != NULL && *(lastSeparator + 1) != '\0') {
+			dirPath = mos_strndup(inputPath, lastSeparator - inputPath + 1);
+			if (!dirPath) goto cleanup;
+            dirPath[lastSeparator - inputPath + 1] = '\0';
+            pattern = mos_strdup(lastSeparator + 1);
+			if (!pattern) goto cleanup;
+            usePattern = TRUE;
+        } else {
+			dirPath = mos_strdup(inputPath);
+			if (!dirPath) goto cleanup;
+        }
+    }
+	
+ 
+
+
+ 	num_dirents = get_num_dirents(dirPath);
+
+
+ 	
+
+
+ 	if (num_dirents == 0) return fr;
+
+
+ 
+
+
+ 	fnos = malloc(sizeof(SmallFilInfo) * num_dirents);
+
+
+ 	
+
+
+ 	fno_num = 0;
+
+
+    fr = f_opendir(&dir, dirPath);
+	
+    if (fr == FR_OK) {
+		
+    printf("Volume: ");
+    if (strlen(str) > 0) {
+        printf("%s", str);
+    } else {
+        printf("<No Volume Label>");
+    }
+    printf("\n\r");		
+		
+	if (strcmp(dirPath, ".") == 0) {
+		f_getcwd(cwd, sizeof(cwd));
+		printf("Directory: %s\r\n\r\n", cwd);
+	} else printf("Directory: %s\r\n\r\n", dirPath);
+		
+        if (usePattern) {
+            fr = f_findfirst(&dir, &filinfo, dirPath, pattern);
+        } else {
+            fr = f_readdir(&dir, &filinfo);
+        }
+
+        while (fr == FR_OK && filinfo.fname[0]) {
+
+	fnos[fno_num].fsize = filinfo.fsize;
+
+
+	fnos[fno_num].fdate = filinfo.fdate;
+
+
+	fnos[fno_num].ftime = filinfo.ftime;
+
+
+	fnos[fno_num].fattrib = filinfo.fattrib;
+
+
+	fnos[fno_num].fname = malloc(strlen(filinfo.fname)+1);
+
+
+	strcpy(fnos[fno_num].fname, filinfo.fname);
+
+
+	fno_num++;
+
+
+
+            if (usePattern) {
+                fr = f_findnext(&dir, &filinfo);
+            } else {
+                fr = f_readdir(&dir, &filinfo);
+            }
+
+            if (!usePattern && filinfo.fname[0] == 0) break;
+        }
+    }
+f_closedir(&dir);
+
+	num_dirents = fno_num;
+	qsort(fnos, num_dirents, sizeof(SmallFilInfo), cmp_filinfo);
+
+
+
+
+
+	if(fr == FR_OK) {
+
+
+		int fno_num = 0;
+
+
+		int col = 0;
+
+
+		for (; fno_num < num_dirents; fno_num++) {
+
+
+			fno = &fnos[fno_num];
+
+
+			if (longListing) {
+
+
+				yr = (fno->fdate & 0xFE00) >>  9;	// Bits 15 to  9, from 1980
+
+
+				mo = (fno->fdate & 0x01E0) >>  5;	// Bits  8 to  5
+
+
+				da = (fno->fdate & 0x001F);			// Bits  4 to  0
+
+
+				hr = (fno->ftime & 0xF800) >> 11;	// Bits 15 to 11
+
+
+				mi = (fno->ftime & 0x07E0) >>  5;	// Bits 10 to  5
+
+
+				
+
+
+				printf("%04d/%02d/%02d\t%02d:%02d %c %*lu %s\n\r", yr + 1980, mo, da, hr, mi, fno->fattrib & AM_DIR ? 'D' : ' ', 8, fno->fsize, fno->fname);
+
+
+			} else {
+
+
+				if (col + strlen(fno->fname) + 1 > scrcols) {
+
+
+					col = 0;
+
+
+					printf("\r\n");
+
+
+				}
+
+
+				if (scrcolours > 2) {
+
+
+					printf("\x11%c%s\x11\x0f ", fno->fattrib & AM_DIR ? 2 : 15, fno->fname);

```

Comment:
explicitly changing colours here is problematic - it changes the current text printing colour and will thus potentially leave the user's terminal set to a different colour than it previously was set to

Reply:
this has been resolved - the function now reads the colours, picks colours accordingly, and reverts the current active colour on completion